catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/security/accesscontrol/TagAggregationVisitor.java

```diff
@@ -0,0 +1,243 @@
+import org.opengis.filter.Filter;
+import org.opengis.filter.Not;
+import org.opengis.filter.Or;
+import org.opengis.filter.PropertyIsEqualTo;
+import org.opengis.filter.PropertyIsLike;
+import org.opengis.filter.PropertyIsNotEqualTo;
+import org.opengis.filter.expression.Expression;
+import org.opengis.filter.expression.Literal;
+import org.opengis.filter.expression.PropertyName;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Visits a filter and aggregates the tags. When the query corresponding to the filter is run, it is
+ * guaranteed any result in the list will have at least one tag found in {@link #getTags()}. If this
+ * cannot be guaranteed, the resultant tag set will be empty.
+ *
+ * <p>Run this visitor against a simplified filter. Running against filters without optimized logic
+ * is not supported. If no other algorithms are available, {@link
+ * org.geotools.filter.visitor.SimplifyingFilterVisitor} can be used.
+ *
+ * <p>Note that {@link PropertyIsNotEqualTo} is not implemented, despite it potentially referencing
+ * metacard tags. That's because aggregation can only effectively operate on what is known to be
+ * true, not necessarily what is known to be false. This is because true predicates narrow down the
+ * result space in deterministic ways without the need of the entire taxonomy or database schema to
+ * compute both sufficient but largely inefficient substitutions.
+ */
+@SuppressWarnings("squid:S1192" /* Constants not actually helpful for log statements */)
+public class TagAggregationVisitor extends TagBaseVisitor {
+  private static final Logger LOGGER = LoggerFactory.getLogger(TagAggregationVisitor.class);
+  private final Deque<TagAggregationRule> ruleDeque = new ArrayDeque<>();
+  private final Set<String> tags = new HashSet<>();
+
+  @Override
+  protected Object preProcessNonTagPredicate(Filter filter, Object data) {
+    reportNonTagEncountered();
+    return data;
+  }
+
+  public Set<String> getTags() {
+    return tags;
+  }
+
+  @Override
+  public Object visit(Not filter, Object data) {
+    ruleDeque.addFirst(TagAggregationRule.newRule());
+    data = super.visit(filter, data);
+    ruleDeque.removeFirst();
+    return data;
+  }
+
+  @Override
+  public Object visit(And filter, Object data) {
+    ruleDeque.addFirst(TagAggregationRule.newRule());
+    data = super.visit(filter, data);
+
+    TagAggregationRule rule = ruleDeque.removeFirst();
+    if (ruleDeque.isEmpty()) {
+      tags.addAll(rule.getTags());
+    } else {
+      ruleDeque.peekFirst().foundTags(rule.getTags());
+    }
+
+    return data;
+  }
+
+  @Override
+  public Object visit(Or filter, Object data) {
+    ruleDeque.addFirst(TagAggregationRule.newRule());
+    data = super.visit(filter, data);
+
+    TagAggregationRule rule = ruleDeque.removeFirst();
+    if (rule.isNotLimitedToTagCriteria()) {
+      // Cannot guarantee result set is a proper subset of policy
+      return data;
+    }
```

Comment:
📖 Note that in this handler for `OR` we are saving the tags **only if** the criteria was limited to tags. That's because, given the nature of an `OR` query, we **can't** assume every result has a tag if the criteria wasn't limited to tags.